### ISSUE TITLE
Indicate in the header what domains are encoded

### ISSUE THREAD

**Issue by lpugin** [music-encoding/music-encoding#318]
Following the numerous and lengthly discussions we had about the encoding of the accidentals, it is pretty clear we need to be able to indicate in the header what domains are encoded in a file in addition to the visual domain. Would `<encodingDesc>` be the right place to look for putting this? Any idea?

**Comment by bwbohl** [comment]
encodingDesc sounds like a reasonable place, at least to me. With projectDesc and samplingDecl it proves to  be designated for such data. Both of which are prose-containers though. Wouldn't a machine readable structure better suit the idea of providing the information about encoded domains?

**Comment by lpugin** [comment]
Yes. We could have an attribute for machine readability and the option to have some text in the element if more information needs to be provided.

**Comment by pe-ro** [comment]
I'd prefer **_not**_ to put such an attribute/element in `<encodingDesc>`. The better place to put the attribute suggested by @kepper at https://github.com/rism-ch/verovio/commit/835d5d37174cdb334611d84ee375cbddb68d7c67#commitcomment-18074646 is on `<score>`.
- Despite the fact that on some level it is metadata (and therefore belongs in the header), it's so closely allied with the musical data that it is more like data itself.
- It can be required on `<score>` without making `<encodingDesc>` required.
- It will be available when the document element is `<score>`.
- Adding it to existing encodings will be simpler.

**Comment by lpugin** [comment]
If we put it on `<score>` I assume we would need it in `<parts>` (and or in `<part>`)

Regarding the requirement and the values, it seems to me that @kepper's suggestion can be seen as equivalent to a `@gestural` that can be "true" (1st case) or "false" (2nd case) and that would be optional (3rd case if not provided). I understand that we do not have the information in existing encodings, hence the 3rd case. At the same time I assume that it it easy to determine cases when it is "true" during the upgrade by looking at the presence of *.ges attributes. As we know, "false" cases cannot be asserted, so the upgrade could not add the attribute. However, as ultimately such an attribute makes really sense if it is mandatory, we can maybe proceed in two steps. A first release with the attribute optional with a warning that it will become mandatory for the next version. This will leave time for upgrading the data. Any thoughts?

**Comment by pe-ro** [comment]
Jo said:

> I believe such a mechanism would be very desirable for other things as well, like indicating how 
> beams have been used (attributes only, always beamSpan, beamSpan only when beam is not 
> sufficient…) etc., or to indicate if the encoding is a transcription of an existing source, or is meant as 
> instruction for a renderer. I'm sure we can come up with better examples, but this mechanism (living 
> somewhere in encodingDesc) could clearly help to make the purpose and meaning of the encoding 
> more explicit.

Ok, but what about those repertoires where `<accid>`, `<beam>`, etc. aren't used at all?  If this "processing information" is required, it will be required in those repertoires too.  How about requiring it in the CMN customization?

**Comment by lpugin** [comment]
I think specifying whereas the gestural domain is encoded or not, or whereas the encoding is a source transcription or not are general concepts not limited to CMN. Providing information about the use of such and such element is probably something different that might need another mechanism.

**Comment by kepper** [comment]
I agree that it's hard to know where which discussion is taking place. I'll just comment here – this time ;-)

A statement like "we considered the gestural domain" is pretty broad-brush – what does it mean to encode the gestural domain? Unless we provide a clear list of which consequences such a statement has for a defined set of attributes / elements, it's almost useless, I believe. Instead, I'd prefer a much more fine-grained mechanism. I don't remember a statement to put something like this on score, but I see the argument. At this point, I'd rather head for either section or scoreDef, as this is also the place where I'd like to see a statement about the repertoire used in a specific section (CMN alternating with Mensural, for instance). Aligning these two things – though they are technically different issues – seems to me like a good idea.

The more I think about it: If we'd come up with a fine-grained mechanism that indicates how certain elements / attributes are used (instead of a statement that the gestural domain has been considered in general), I would put that in an element of it's own.This element could go into scoreDef and / or section, but could equally well live inside encodingDesc (because this place just _feels_ right). In that case, scoreDefs and / or sections could just point to it using @decls. If somebody splits up his file and starts with score as document element, then he can still use an anyURI inside @decls to resolve that correctly.

Comments?
jo

**Comment by lpugin** [comment]
> A statement like "we considered the gestural domain" is pretty broad-brush – what does it mean to encode the gestural domain? Unless we provide a clear list of which consequences such a statement has for a defined set of attributes / elements, it's almost useless, I believe

I agree we need good explanations for a `@gestural` attribute (or whatever it is called), which I guess should go in the guidelines. I am volunteering to do.

>  I'd rather head for either section or scoreDef, 

I like it better than score too.

Having a fine grained description about the use of elements would require its own element(s). I also feel that those should go in encodingDesc.

**Comment by craigsapp** [comment]
> A statement like "we considered the gestural domain" is pretty broad-brush – what does it mean to encode the gestural domain? Unless we provide a clear list of which consequences such a statement has for a defined set of attributes / elements, it's almost useless, I believe.

Yes.  There can be many gestural components, and maybe only some are encoded and others are not, so it is probably wisest to specify which of the individual components are considered.  And possibly a code indicating that all of them as a group have been considered.

The two that I have had direct experience with so far in converting from Humdrum into MEI are:
(1) note@accid.ges 
(2) note@oct.ges

note@oct.ges is used when the note is being affected by an ottava marker.  

I haven't dealt with transposing parts yet, but in theory this could be another gestural domain category.  Probably you do not want to give the sounding pitch as the gestural information for the note element in a transposed score.  But perhaps a machine-readable code asserting that transposition instructions are embedded for the parts which can be use to generate sounding pitches would be good.

# 

The problem can also be further generalized beyond "gestural", so you might want to think more before taking any action.  The core of the problem is how to interpret the meaning of unspecified attributes.  Either the unspecified attribute means "undefined" or it means "use a default value".  

Or another way of describing it, slightly more formalized and generalized: 
- CONDITION 1: "use a default value" means: "generate a value which can be calculated from information within the current element and its sub-elements"
- CONDITION 2: "undefined" means: "generate a value which must be calculated by looking outside of the current element into the wider XML tree structure".   

Computational generation of unspecified attributes must be given information as to which of these two conditions should be applied, since both cannot be used at the same time (given trivial caveats of checking sub-elements for explicit information before heading out to the wider XML structure to generate an implicit value).

Example for accidentals:

```
<note oct="4" pname="f"\>
```

You assert that not specifying @accid in this situation means @accid="", which is a CONDITION 1 decision.  As this is the only interpretation allowed, there does not have to be an indication that CONDITION 1 is being applied.

But for generating @accid.ges, you apply CONDITION 1 or CONDITION 2 as suits your needs.  Therefore, which of the two cases is to be used must be specified, as the computer has no idea itself which to use.  For many cases it could be inferred, but not all cases, so the method must be made explicit.

For CONDITION 1, @accid.ges="n" because there is no @accid, so the algorithm chooses a default of being natural.

For CONDITION 2, @accid.ges is locally undefined.  It must be defined by potentially reading the entire XML tree structure representing time before the current point in the score.  This is easy for musicians to do, but is quite complex to do properly in a computer algorithm.  It is solvable to a reasonable degree of accuracy, but is a pain to have to deal with.

Here is a basic case:

```
<layer n="1">
<note oct="4" pname="f" accid="s" dur="4">
<note oct="4" pname="f" dur="4">
</layer>
```

For the second note, using CONDITION 1 will return a value of @accid.ges="n", but CONDITION 2 will return a value of @accid.ges="s" since the note played previously in the layer has been sharpened.

Here is a more complicated example:

```
<layer n="1">
<note oct="4" pname="f" accid="s" dur="4">
<note oct="4" pname="f" dur="4">
<layer n="2">
<note oct="4" pname="g" dur="8">
<note oct="4" pname="f" accid="n" dur="8">
<note oct="4" pname="e" dur="8">
<note oct="4" pname="d" dur="8">
</layer>
```

For the second note, using CONDITION 1 will return a value of @accid.ges="n".  This is the same as the previous example, and will always be true because the external context of the note is irrelevant.  However for CONDITION 2, @accid.ges="n" instead of sharp.   This is due to the @accid="n" from the second note in layer 2.

CONDITION 2 generation of accid.ges involves traversing the entire XML tree.  In particular the worse-case situation involves notes which are placed onto other staves other than the home staff for the given layer.  There are also syntactic rules for grace notes versus regular note accidentals, as well as more commonly dealt with barlines and key signatures.

And I will also mention as an aside that orchestral scores occasionally use a slightly different algorithm for determining gestural accidentals.  If two instrumental parts are displayed on a single staff, their accidentals do not influence each other, so:

```
<layer n="1">
<note oct="4" pname="f" accid="s" dur="4">
<note oct="4" pname="f" dur="4">
<layer n="2">
<note oct="4" pname="g" dur="8">
<note oct="4" pname="f" accid="n" dur="8">
<note oct="4" pname="e" dur="8">
<note oct="4" pname="d" dur="8">
</layer>
```

In this example, @accid.ges="s" for the second note instead of "n" as assigned by the standard algorithm.  If you want to be 100% precise in how the MEI data should be interpred, you would need to also encode the algorithm which should be used to calculate @accid.ges, which in this case would be "partOverlay" or something to indicate a non-standard algorithm (which might mean call the standard algorithm "standard", and there would be an undefined case of "unknown" if you truly don't know how to interpret the gestural accidentals from the printed score.).

**Comment by pe-ro** [comment]
Isn't this a very complex way of requiring `@accid.ges` on a note that doesn't have `@accid`?

**Comment by lpugin** [comment]
I am not sure it would make much sense to encode accid.ges but not oct.ges. But you are right that if you want to do so and have a coded way to say "I have encoded accid.ges but not oct.ges" than we need more than one attribute. Of course, this is the freedom of MEI and nobody can force you to encode anything. But if nobody wants to give up a little bit of his freedom, then adding a pile of optional elements and attributes will just be meta-coding of the ambiguity and create some more ambiguity.

Maybe we need to think differently, since we seem quite far from a consensus, on the fact that this is actually needed, that it might be required, or how to do it. As @pe-ro said, it is ambiguous. We can still clarify the guidelines. MEI is un-ambiguous in the visual domain, and what is encoded is what you see. Then for the rest, it depends. Some encode accid.ges, some do not, some encode accid.ges="undefined", some do not, some use ploc and oloc, some use pname and oct. You need to be aware of this, and depending on what you want to do your software will need to deal with this. Wouldn't this be clearer? I think it would.

**Comment by pe-ro** [comment]
@lpugin, I'm not clear about what you're suggesting -- What would be clearer?  Providing more explanation in the Guidelines?

**Comment by pe-ro** [comment]
To test the answer to my last question, here are Craig's examples with explicitly encoded `@accid.ges` attributes where necessary.

The basic case.

```
<layer xmlns="http://www.music-encoding.org/ns/mei" n="1">
  <note oct="4" pname="f" accid="s" dur="4"/>
  <note oct="4" pname="f" dur="4" accid.ges="s"/>
</layer>
```

A more complicated example.

```
<staff xmlns="http://www.music-encoding.org/ns/mei">
  <layer xmlns="http://www.music-encoding.org/ns/mei" n="1">
    <note oct="4" pname="f" accid="s" dur="4"/>
    <note oct="4" pname="f" dur="4" accid.ges="s"/>
  </layer>
  <layer n="2">
    <note oct="4" pname="g" dur="8" accid.ges="n"/>
    <note oct="4" pname="f" accid="n" dur="8"/>
    <note oct="4" pname="e" dur="8" accid.ges="n"/>
    <note oct="4" pname="d" dur="8" accid.ges="n"/>
  </layer>
</staff>
```

Orchestral example.

```
<staff xmlns="http://www.music-encoding.org/ns/mei">
  <layer n="1">
    <note oct="4" pname="f" accid="s" dur="4"/>
    <note oct="4" pname="f" dur="4" accid.ges="s"/>
  </layer>
  <layer n="2">
    <note oct="4" pname="g" dur="8" accid.ges="n"/>
    <note oct="4" pname="f" accid="n" dur="8"/>
    <note oct="4" pname="e" dur="8" accid.ges="n"/>
    <note oct="4" pname="d" dur="8" accid.ges="n"/>
  </layer>
</staff>
```

In no case can its value be prescribed (that's context dependent), but requiring `@accid.ges` (via Schematron) is a fairly simple solution to the problem, isn't it?

Doing so doesn't violate our earlier decision to require that `@accid` and `@accid.ges` never have the same value.

**Comment by lpugin** [comment]
Requiring `@accid.ges` via Schematron would break compatibility with current encodings, so I think this is a no go.

**Comment by pe-ro** [comment]
As @lpugin suggested earlier, we could create a Schematron rule in the next release that warns of missing `@accid.ges` attributes and states that the following release will require them.

**Comment by pe-ro** [comment]
Another possibility is to mark `@accid.ges` as "recommended" as proposed over in #320.  This would be accompanied by an explanation of why it's a good idea to use this attribute if you're interested in analysis and/or automated performance.

**Comment by pe-ro** [comment]
> Requiring @accid.ges via Schematron would break compatibility with current encodings, so I think this 
> is a no go.

We could phase it in with a warning first as you suggested earlier.  We could also allow a value of "unknown", "default", "do-what-you-will", etc. which could be applied to existing encodings until the next-next release.

**Comment by pe-ro** [comment]
If we proceed with this idea, it might be a good idea to rename `@accid.ges` to `@accid.default` (that is, effectively move it out of the gestural domain and into the logical) so that it won't be removed if/when the gestural domain is excluded by a customization.

**Comment by lpugin** [comment]
If it is possible, then I like requiring `accid.ges`, even though it is more constraining than requiring an indication in the header. I am not sure about the renaming. What about `@accid.log`?

**Comment by pe-ro** [comment]
`@accid.log` sounds good.

Just to make sure we're talking about the same thing, I meant that Schematron would be used to enforce a rule that notes without `@accid` would have to have `@accid.ges`.  Are we agreed?

Can you explain why is this more constraining than requiring an indication in the header?

**Comment by lpugin** [comment]
> Just to make sure we're talking about the same thing, I meant that Schematron would be used to enforce a rule that notes without @accid would have to have @accid.ges. Are we agreed?

Yes, this is what I understood

> Can you explain why is this more constraining than requiring an indication in the header?

Because the solution with the header (something like `@ges="false"`) makes it possible not to encode the gestural domain at all. But I think your proposal is better.

**Comment by kepper** [comment]
I'm not sure I'm willing to follow here. We're not enforcing things like dur, pname and oct, and now we're asking to provide accid.log="n" on every single note out there that doesn't ask for something else? Seriously?

If I got that comment right (and maybe I didn't), it's not more constraining, it's more demanding. Having a more or less simple statement in one place is one thing, polluting every single MEI file with an additional required attribute on every note is definitely another thing. Especially when we consider how long we managed to live with the current situation, and that this issue really only affects one specific use case for MEI. I'm not saying we should stick with the current solution – there is definitely room and / or need for an improvement, but this approach seems to be over the top in my eyes. Why do we force projects to care about the gestural domain when they really don't. Why can't they say just that? In some sense, this feels like TEI would force everyone to mark up individual syllables… (I'm deliberately exaggerating)

**Comment by pe-ro** [comment]
Here's what I've got so far for the Schematron rule --

```
<pattern>
  <sch:rule context="mei:note[@accid]">
    <sch:let name="this.pname" value="@pname"/>
    <sch:let name="this.oct" value="@oct"/>
    <sch:assert test="not(following-sibling::mei:note[@pname=$this.pname 
      and @oct=$this.oct and not(@accid)][not(@accid.ges)])">When a note 
      has a written accidental (@accid), all following notes in the same 
      layer with the same pitch name and octave number must have 
      a gestural accidental (@accid.ges).</sch:assert>
  </sch:rule>
</pattern>
```

The following fragment will fail validation because the 2nd note in layer 1 doesn't specify its gestural accidental.

```
<staff xmlns="http://www.music-encoding.org/ns/mei">
  <layer n="1">
    <note oct="4" pname="f" accid="s" dur="4"/>
    <note oct="4" pname="f" dur="4"/>
  </layer>
  <layer n="2">
    <note oct="4" pname="g" dur="8"/>
    <note oct="4" pname="f" accid="n" dur="8"/>
    <note oct="4" pname="e" dur="8"/>
    <note oct="4" pname="d" dur="8"/>
  </layer>
</staff>
```

There's no way to determine whether the value of `@accid.ges` on the 2nd note should be 's' or 'n', but least we can draw the encoder's attention to the ambiguity.

**Comment by kepper** [comment]
the problem is that this Schematron matches current practice, but fails for older users of CMN. Even Johann Sebastian Bach sometimes used the old model, which just says that accidentals apply as long as the same pitch is repeated – even across barlines, and needs to be re-set after one different pitch. If you have c# c c c | c c c c, all notes are played as c sharp. If you have c# d c c …, only the first is a c sharp, while all following c's are just that. This is not a corner case, and the bad thing is that quite frequently, you'll find both systems side by side in the same score. Of course it's possible to write Schematron rules for each one of them (and maybe we should offer them as separate schemas, to be applied by the user), but we can't come up with a combined schema that addresses both at the same time – they're just contradicting, and it needs someone to identify the current ruleset before we can resolve them. In the Freischütz, it was easy to add all accid.ges automatically by XSLT since we knew the rules stayed the same throughout the project. On the level of MEI in general, I don't believe we can do the same. 

Re: the 's' and 'n' question: Yes you could decide which value it should be. Just store the current accid value in another variable and compare each subsequent note with that… But again, this should happen in a separate Schematron that people may or may not add to their files, depending on whether it makes sense for their encodings or not…

**Comment by lpugin** [comment]
It seems indeed that it will be difficult (if possible at all) to have a Schematron rule to deal with various rules. What I understood from your idea @pe-ro would be that `@accid.ges` would be required in any case.

So maybe we need both: a value in the header that says that the gestural domain is encoded, and then this triggers the Schematron rule that just checks if `@accid` and / or `@accid.ges` is provided. If you are not interested by the gestural domain and do not want to pollute your file, you turn it off and then `@accid.ges` is not allowed (by Schematron).

**Comment by lpugin** [comment]
> Especially when we consider how long we managed to live with the current situation, 

It does not mean that it is ideal ;-) And maybe because we have not shared that much data?

> and that this issue really only affects one specific use case for MEI.

I am not so sure about it. Actually one reason why we are having this discussion is because MEI has taken a  kind of half-baked approach regarding visual vs gestural domain. SCORE has a purely visual domain approach. A note is encoded with it line position (basically the equivalent of oclo and ploc). Now in MEI we generally use pname and oct, which was quite likely chosen because it facilitates data reading - human reading, but also machine (e.g., analysis) reading.  This is clearly leaning towards the gestural domain. So in a way we cook it for the clefs, but then remain pretty unclear whereas the key signatures were also cooked or not. (When MEI is rendered it is necessary to process back the clef but not the key signatures. This illustrates in the other direction this intermediate position in which MEI has put itself.)

> there is definitely room and / or need for an improvement, but this approach seems to be over the top in my eyes.

I guess we agree, then. What would be your proposal for improvement?

**Comment by kepper** [comment]
I would
- add something as a general indicator how the gestural domain has been treated
- this should allow to separate between accid.ges, transposing instruments etc.
- this mechanism should be repeatable in the header and addressable from section and / or scoreDef to allow mixing of encodings of different origin
- this mechanism _must_ offer an option like "undetermined" for older encodings. If we like, we can deprecate this value right from the beginning and announce that it'll go away in reasonable time
- accept your offer to write a chapter for the guidelines, and doing so hope to sensitize encoders to do it right (TM)

I would not
- force everyone to care about the gestural domain – some people just don't
- require accid.ges even for "neutral" notes which aren't affected by anything, i.e. no rule that either accid or accid.ges have to be present
- change the name of accid.ges to accid.log -> this could be too confusing for existing users
- bake a Schematron rule _directly_ into MEI to check for correct use of accidentals (but offer them as optional "helpful additional schemas" instead)
- try to take away all (technical) ambiguity by validation
- bother if our final solution violates existing MEI versions, as long as we provide and document a method to bring those older instances to the newer version (this argument reminds of some other music encoding scheme, and we all know where this ended…).

I believe that something along these lines balances the required additional encoding efforts for every encoder and the software developer's convenience when reading those files. Yes, we as developers still need to have some trust that the encodings are correct, but for years it has been consensus that we can't keep people from doing dumb stuff with MEI. We can try to provide better guidance on how to prevent that, both with technical means and better explanation, but we should not lock down everything to prevent errors in the first place. I believe we will loose too much when taking this approach. We should strongly encourage people to either use accid.ges (and similar) consequently and correctly (so that software can rely on it) or not at all – in which case software has to fill it up according to a ruleset the developer thinks is appropriate. Actually, for the latter practice in CMN, I've written an XSLT that automatically adds accid.ges to all affected notes (including notes tied across barlines, which haven't been covered in @pe-ro's schematron yet). It needs some brushing up, but could then be offered as generic tool to add accid.ges automatically. However, it is up the user to decide If the rules in there apply, or if something else is needed.

Just my 2c,
jo

**Comment by lpugin** [comment]
> - this should allow to separate between accid.ges, transposing instruments etc.

I cannot see where it would make sense to have accid.ges but not transposing instruments or not oct.ges. But never mind, how would you do it? So could you elaborate? As far as I understood @craigsapp would like to have this too.

**Comment by kepper** [comment]
Of course I haven't thought much about the name of the element, but why not something like 
`<supported accid.ges="yes" trans.semi="no" trans.diat="no" oct.ges="unknown"/>`? This may cause problems where we have multiple definitions for attributes of the same name (like on `@dur`), but I'm not sure if these overlap with the problem at hand. Of course it'll need thorough documentation what this means, but I heard an excellent writer already volunteered for covering that ;-)

**Comment by kepper** [comment]
In order to avoid those double definitions in the future, it might actually be better to prefix them in this element, like `<supported supports.accid.ges="no"/>` or some such…

**Comment by kepper** [comment]
maybe the element could then be name as `<encoding supports.accid.ges="unknown"/>` or encodingStyle, encodingPractice, encodingGuidelines etc. …

**Comment by pe-ro** [comment]
@kepper  said:

> I'm not sure I'm willing to follow here. We're not enforcing things like dur, pname and oct, and now we're 
> asking to provide accid.log="n" on every single note out there that doesn't ask for something else? 
> Seriously?

I think you misunderstood my schematron suggestion -- 
1. it only affects those notes that are preceded by a note with a written accidental, and
2. it can issue a warning, which the processor is free to ignore, instead of raising an error.

Perhaps we need to investigate the possibility of conditionally activating Schematron rules; that is, by placing them in their own modules, for example MEI.cmn.bach-accid and MEI.cmn.weber-accid.

> offer them as optional "helpful additional schemas" instead

This sounds a lot like customization. :-)  Which I suggested several days ago, BTW.

**Comment by pe-ro** [comment]
> MEI has taken a kind of half-baked approach regarding visual vs gestural domain

Ouch!

**Comment by lpugin** [comment]
> I think you misunderstood my schematron suggestion -- 
> 1. it only affects those notes that are preceded by a note with a written accidental, and

I understood it as @kepper (having `@accid.ges` on every note). So then I agree that defining the rule will be too difficult because of the different practices.

I think @kepper's suggestion with an element and distinct attribute is a way to go. I wonder how many more attribute we would have. I also wonder if this is fine grained enough. For example, you can encode `@dur.ges` on notes but not on dynamics and want to specify this.

**Comment by lpugin** [comment]
@pe-ro half-cooked can be good. At least certainly for foie gras... ;-)

**Comment by pe-ro** [comment]
> I understood it as @kepper (having @accid.ges on every note). So then I agree that defining the rule 
> will be too difficult because of the different practices.

That just means we need different rules for different situations.

> I think @kepper's suggestion with an element and distinct attribute is a way to go. I wonder how 
> many more attribute we would have. I also wonder if this is fine grained enough. For example, you 
> can encode @dur.ges on notes but not on dynamics and want to specify this.

That's why this approach is problematic -- taken to its logical end, it means replicating the schema in a sense.  If we're going toward that, then why not allow users to put a customization file directly in the header as TEI does?  In combination with the schema, this would provide all the info one needs regarding what is/isn't supported.

**Comment by craigsapp** [comment]
@kepper wrote:

> We can try to provide better guidance on how to prevent that, both with technical means and better explanation, but we should not lock down everything to prevent errors in the first place.

There is no need to lock it down (but that would be nice).  The sufficient minimum that I want is a marker in the header which says @gestural="true|false" which means that the gestural information (information needed for MIDI rendering and musical analysis) has or has not been encoded in the data.  A MIDI renderer should refuse to process data having @gestural="false" by itself unless explicitly instructed on how to calculate @accid.ges (@oct.ges, transpositions, etc).

The marker could be optional for a while, then recommended, then should be required eventually.

Regarding locking down, I would say that

```
<note pname="f" />
```

should be encoded as

```
<note pname="f" accid="none" />
```

The first encoding implies that the state of @accid is not important, but you seem to place a much greater importance on there being no visible accidental with this encoding.  Not specifying anything for @accid implies that software processing can do what it likes with that attribute.

**Comment by lpugin** [comment]
The risk of having this becoming too complex is also the reason why I suggested a simple attribute. Basically to say, "yes, gestural attributes are encoded and this file can be used as is for analytical purposes". What this means indeed needs to be clarified, but we can think of the attributes that are necessary to produce a "correct" MIDI file. So far we had `accid.ges`, `oct.ges`, `trans.diat`, `trans.semi`. Maybe a few more, but probably not many.

The fine grained encoding description is another story. It actually also applies to the visual domain (e.g., are stem dir encoded?). And this can indeed duplicate the role of a customization. So we do not need to solve everything in one go.

**Comment by kepper** [comment]
>  taken to its logical end, it means replicating the schema in a sense.

but, when we decide to use different attribute names, with the possibility to group things. There's no need for `@supports.trans.diat` and `@support.trans.semi` side by side – `@supports.transpositions` will cover both attributes. At this point, we've only spotted octave displacement, transposing instruments and sounding accidentals as areas of interest for this. That's not really replicating the schema, is it? However, I believe that transposing instruments are different from the two other things, and having three attributes instead of one is a price I'd be willing to pay given that we can easily extend this mechanism in the future if need be…

> If we're going toward that, then why not allow users to put a customization file directly in the header as TEI does?

Maybe I'd prefer to just reference it, instead of putting it there, but that's splitting hairs. Ignoring these implementation details, I'm all in for this. It's just a solution for a different problem, which only partly affects the current issue. How many projects actually do remove unsupported attributes from their schema? Relying on that doesn't seem to be any safer than the current situation…

**Comment by pe-ro** [comment]
> How many projects actually do remove unsupported attributes from their schema? 

Probably none, but that would be the first step in the creation/use of a schema-aware processor.

**Comment by pe-ro** [comment]
> The sufficient minimum that I want is a marker in the header which says @gestural="true|false" which 
> means that the gestural information (information needed for MIDI rendering and musical analysis) has 
> or has not been encoded in the data. A MIDI renderer should refuse to process data having 
> @gestural="false" by itself unless explicitly instructed on how to calculate @accid.ges (@oct.ges, 
> transpositions, etc).

In the header, or on `<score>` (and by extension, on `<part>` as well)?

**Comment by pe-ro** [comment]
I know that maximum interoperability calls for making `@gestural`, etc. required, but could it be optional? MIDI renderers would be expected to process files with `@gestural="true"`, refuse to process files with `@gestural="false"`, and decide what to do when `@gestural` isn't present.

**Comment by lpugin** [comment]
I think a gentle transition in two steps (optional with a deprecation warning and then required) would be better. It is actually a good opportunity to draw people's attention on this. What does it take to think if it has to be true or false? You probably have to do it once for your project, or your dataset, and that is it. For the community ;-)

**Comment by craigsapp** [comment]
@kepper wrote:

>  In the Freischütz, it was easy to add all accid.ges automatically by XSLT since we knew the rules stayed the same throughout the project. On the level of MEI in general, I don't believe we can do the same.

If you, the encoder of the Freischütz data, don't add @accid.ges, who will?  And will they do it correctly?  And would you otherwise quantify the conversion rules in the data? This seems to be an argument that @accid.ges be required.

If an encoder does not know how to calculate @accid.ges, then they should say so explicitly, and give an explanatory note in `<encodingDesc>`. Late Medieval and early Renaissance repertories are the only which can have an excuse to not place @accid.ges in the data.  Usually the composer did not write accidentals, but expected the singers to use them when performing.   In this case, a careful reading of the manuscript cannot add @accid.ges, since it is undefined in the score and convention of the day.  Starting around the mid-16th century composers had to start writing accidentals because they were writing in a new style which increasing modulations, and singers could not get the implicit accidental system to work any longer.  So from the 17th century on, musicians expect accidentals to be fully specified in the music, and there should not be a problem interpreting @accid.ges from a score, given its historical context.  If the conversion process is not according to modern rules, either the @accid.ges should be provided by the encoder, or some automatically readable mark as to what convention is being used should be given.

> Actually, for the latter practice in CMN, I've written an XSLT that automatically adds accid.ges to all affected notes (including notes tied across barlines, which haven't been covered in @pe-ro's schematron yet). It needs some brushing up, but could then be offered as generic tool to add accid.ges automatically. However, it is up the user to decide If the rules in there apply, or if something else is needed.

Components of converting between @accid and @accid.ges (in standard modern conventions):
(1) key signature
(2) previous occurrence of diatonic pitch in measure
(3) previous occurrence of pitch in previous measure if tied over
(4) grace notes take state of previous normal pitch in measure or immediately preceding grace notes in group.
(5) normal notes are not influenced by grace notes accidental states
(6) cross-staffed notes take the context of the staff on which they are visible and do not affect their home staff notes.

You still have cases 4, 5 and 6 to implement I presume :-), and any others that I have not thought of...

Here is the sample of cross-staff notes/stems/beams, which illustrate case (6) (arrows mark notes which are affected by the rule):

![screen shot 2016-07-01 at 7 27 02 am](https://cloud.githubusercontent.com/assets/3487289/16525921/8d50d62a-3f63-11e6-9c3b-a74a8d843b36.png)

It is not a problem going from visual to gestural accidentals, as long as the convention to do so is defined, so from a software point of view it is not too big a deal.  The main problem at the moment is @accid.ges is an orphan child in MEI due to the current user base being strongly visually oriented.  Now that there is a MIDI exporter in verovio for MEI data, there will be more need for proper handling of performance gestural information, just as there was after the invention of the standard MIDI file.  SCORE data can be used to generate MIDI files, and Humdrum data can be used to print scores.  They are very clear on what is being represented, so it is not much of an issue to use either in both usage cases, although they are optimized for a specific case.

Note that most musical codes generated after the development of MIDI encode gestural accidentals, while pre-MIDI representations encode visual accidentals.  This is why calling @accid @accid is bad as it is an arcane viewpoint, and it would better be called @accid.vis, as most any other data format currently uses @accid as the name of @accid.ges.  

MuseData has a double-encoding system for accidentals which was adopted by MusicXML.  Accidentals are primary encoded in gestural, and the visual are usually extracted from the gestural, but can be encoded explicitly for unusual situations, such as the case where a sharp means a natural.  In that case the gestural would be natural and the visual would be sharp.  

MEI can do the exact opposite without problems if done systematically, but fluffing off of gestural accidentals is not good.  The gestural accidentals are needed for manipulating the music.  For example, if you want to create a modern performing edition of Freischütz, you are not going to use early 19th century rules, are you?  Even if it is a modern critical edition, you would most likely put editorial accidentals on non-conventional notes, which is basically encoding @accid.ges in the visual domain.

**Comment by kepper** [comment]
Who said that we haven't considered `@accid.ges`? Of course we did – even though we didn't need them in the project itself. My converter may not be perfect for every case, but it did for all cases we had in the Freischütz, and it may serve as a starting point for an improved version. If you prefer to write something on your own, which takes care of other corner cases, feel free to do so. Axel Berndt also wrote something along these lines, with some MEI guidance by @bwbohl. 
However, I still refuse to accept that we enforce the encoding of `@accid.ges`. Any _naive encoder_ (as @pe-ro calls them) may have a problem to identify this, and we should allow for a second step to improve the data into that direction, instead of requiring it in the first place. I don't buy that the gestural domain is more important than the visual – they are both equally important from an MEI perspective (even though only one of them can get them name without the added .ges or .vis), but may be of varying importance to any given project. Forcing everyone to do certain things just because these things are relatively important in your world doesn't seem to be right. 

just my 2c

**Comment by lpugin** [comment]
At this stage I would like to make a proposal because we have discussed the topic quite extensively.

Since it seems to be unacceptable to require the encoding of `accid.ges` (and of the gestural domain in general), then we need to limit ourselves to providing information whereas it is encoded or not. It is acceptable to ask users to inform others about it, so my proposal is:
- to make a `@ges="yes|no"` required in two steps (optional with a deprecation warning and then required). 
- to put it in `<scoreDef>`
- to have a Schematron rule to require it only in the initial `<scoreDef>` that appears at the beginning (traditionally before the `<section>` element - then it should be optional for signaling a change, even though it will be rarely the case.
- to have a Schematron rule that checks that `@accid.ges` is not used when `@ges="no"`. 
- to have a Schematron rule that checks that `@accid.ges="n"` never occurs without `@accid`

Of course there is always the possibility for a project to change this with a customization.

How does it sound?

**Comment by zolaemil** [comment]
To me this sounds like a customization problem. Projects that rely on the use of `@accid.ges` could use a customization which makes it mandatory.

**Comment by lpugin** [comment]
How can you make it required by a customization? As explained above by @kepper, it would mean having Schematron rules that take into account accidental music notation rules, which indeed seem impossible to define. Another way to look at this: if you have a piece in C-major with no modulation, they will be no `@accid.ges` to require.

**Comment by zolaemil** [comment]
ah sorry, it did seem to easy to be true... :)

**Comment by lpugin** [comment]
And another way to look at this: the visual domain is primary in MEI because need to choose one, and this is fine. But it still is an analytical encoding and not a purely visual encoding. To encode this
![image](https://cloud.githubusercontent.com/assets/689412/16615261/a7802b34-4375-11e6-9ef5-183858055b8e.png)
 in a purely visual manner (e.g., alla SCORE), you would do:

```
<note loc="4" dur="4"/>
```

Now in MEI, assuming a G treble clef, we usually do:

```
<note pname="b" oct="4" dur="4"/>
```

So we do the diatonic analytical part of the processing (reading the clef), but not the chromatic part (reading the key signature or previous accidentals) when `@accid.ges` is omitted. This is what I mean by half-baked. For me it is problematic but only because we have no indication about it. If you are told that `@accid.ges` is omitted, you can try to do something, or exclude the file. But if not, then this is what you will get - please listen to it.
http://www.verovio.org/mei-viewer.xhtml?file=examples/ges/Bach_Wie_bist_du_Seele.mei

Happy with this?

**Comment by kepper** [comment]
I just met with @lpugin, and the two of us reached something that we consider as a pragmatic compromise of all things that have been discussed. It consists of a mandatory part, and an additional part that we'd like to poll users on MEI-L whether we actually need it or not. We still haven't discussed the implementation details like element / attribute names, so please consider all this as preliminary. Here's the plan: 
- in the next release, require an attribute called `@gestural` on the first instance of `<scoreDef>` in a `<score>` or `<part>` (subsequent instances are optional, if not provided, the previous value is kept).
- allow the following values: _true_, _false_, _undetermined_. Out of these, _undetermined_ will be deprecated right away, and will go away in the following release. This gives projects time for transition, but it enforces a statement of some kind immediately. In essence, we don't have to guess what the absence of `@gestural` means. 

This is the most simple solution to determine whether the gestural domain has been considered or not in an encoding – there's no guessing what the absence of corresponding attributes means. It requires thorough explanation in the Guidelines, but technically, it's bloody simple and is certainly not too demanding. 

However, it only allows an "all-or-nothing" kind of encoding. At the same time, we're uncertain too which degree projects so far supported the gestural domain only partly – perhaps by adding info about transposing instruments, but not `@accid.ges`. If there is a significant number of encodings out there which is in this "half-baked" state, the solution above will either force them to add additional information and thus upgrade their encodings, or to remove parts of the information they already have (considering that we use this mechanism to trigger some Schematron rules and they don't customize the whole thing like they should). This is the point where we'd like to poll MEI-L and ask how they've dealt with the gestural domain in the past. This poll will result in one of two options:
1. (Almost) everyone has used the gestural domain either completely or not at all. In this case, we can stop it and stick with the solution above. 
2. There is indeed a significant amount of encodings with partial support of the gestural domain. If this is the case, then we'd like to add the following points to our proposal above:
- add another value called _mixed_ to `@gestural`. 
- this value will trigger the need for a small set of additional and more specific attributes on `<scoreDef>`, like `@gestural.events`, `@gestural.trans`, `@gestural.controlEvents` or some such. Each of these will have the values _true_, _false_ and _undetermined_ (which will also be deprecated and will go away in the next-next release). 
- if `@gestural` is set to _mixed_, all of these additional attributes will be required
- The guidelines will clearly specify what each value for these attributes means. We will follow the current use of "half-baked" encodings to come up with the smallest possible number of still rather broad-brush categories (`@gestural.events` will cover both `@accid.ges` and `@oct.ges`, for instance). 
- We can work on those implementation details of this finer-grained mechanism after the poll. 

This additional and complementary mechanism will allow us to be more specific than the one-size-fits-all solution above, and builts in some kind of extension mechanism. However, it is not meant as a kind of secondary customization mechanism, nor does it seek to reference an ODD file. It's really more like a formalized encoding description. 

We also discussed to put this in the header, preferably in `<encodingDesc>`. In case we (=MEI) want to do this as well, we (= @lpugin and me) would like to do it as follows:
- have an element somewhere inside `<encodingDesc>`. It should be repeatable in order to mix encodings from different sources, which may actually follow different practices. 
- The element should offer the same set of attributes as specified above for `<scoreDef>`. Maybe, these attributes could also be made available as child elements to allow specific explanations or each of those aspects. 
- There should be room for additional explanations on the encoding guidelines used for generating these instances. 
- In order to link from an actual ecoding to these definitions, we propose an additional value on `@gestural` called _external_ (or similar). which will trigger a (then) required attribute `@gestural.def` of type _anyURI_. This may point to the corresponding element in the header, or to some generic project-level definition outside the file. 

However, this mechanism of describing the support for the gestural domain in the header is mostly indepent from our proposal above. What we'd like to hear now is if we all can agree on the basic mechanism described at the top, and the procedure to query MEI-L for the necessity of the finer-grained solution described thereafter. If there are no timely complaints about this, we will approach MEI-L by the end of this week. As soon as we have an answer to that, we can start to discuss and work out the implementation details for this part (if needed). 

All best, Laurent and Johannes

**Comment by pe-ro** [comment]
I'm not completely opposed to these ideas; however, there may be more than one way to implement them. For the sake of maximum explicitness, expressing as many constraints as possible in the schema itself is often the best way to go.  But when there are competing purposes for which the markup must be explicit, then it's not possible to put every constraint in the schema.  Expressing contextual constraints, of course, is the _raison d'etre_ for Schematron.

We're well past the point where the MEI schema is meaningful without the accompanying Schematron rules, so I offer the following as an alternative to making changes to the schema --

```
<pattern xmlns="http://purl.oclc.org/dsdl/schematron">
  <!-- key signature: sharp keys -->
  <sch:rule xmlns:rng="http://relaxng.org/ns/structure/1.0"
    xmlns:sch="http://purl.oclc.org/dsdl/schematron"
    xmlns:a="http://relaxng.org/ns/compatibility/annotations/1.0"
    xmlns="http://www.tei-c.org/ns/1.0" xmlns:math="http://exslt.org/math"
    context="mei:note[not(@accid) and //mei:*[@accid.ges] and
    contains (preceding::mei:*[local-name()='staffDef' or local-name()='scoreDef'][@key.sig][1]/@key.sig, 's') and 
    contains(substring('fcgdaeb', 1, number(replace(preceding::mei:*[local-name()='staffDef' or local-name()='scoreDef'][@key.sig][1]/@key.sig, '[s]', ''))), @pname)]">
    <sch:assert role="warning" test="@accid.ges">keysig: <sch:value-of
      select="preceding::mei:*[local-name()='staffDef' or local-name()='scoreDef'][@key.sig][1]/@key.sig"
      />, @accid.ges required.</sch:assert>
  </sch:rule>
</pattern>
<pattern xmlns="http://purl.oclc.org/dsdl/schematron">
  <!-- key signature: flat keys -->
  <sch:rule xmlns:rng="http://relaxng.org/ns/structure/1.0"
    xmlns:sch="http://purl.oclc.org/dsdl/schematron"
    xmlns:a="http://relaxng.org/ns/compatibility/annotations/1.0"
    xmlns="http://www.tei-c.org/ns/1.0" xmlns:math="http://exslt.org/math"
    context="mei:note[not(@accid) and //mei:*[@accid.ges] and 
    contains (preceding::mei:*[local-name()='staffDef' or local-name()='scoreDef'][@key.sig][1]/@key.sig, 'f') and
    contains(substring('beadgcf', 1, number(replace(preceding::mei:*[local-name()='staffDef' or local-name()='scoreDef'][@key.sig][1]/@key.sig, '[f]', ''))), @pname)]">
    <sch:assert role="warning" test="@accid.ges">keysig: <sch:value-of
      select="preceding::mei:*[local-name()='staffDef' or local-name()='scoreDef'][@key.sig][1]/@key.sig"
      />, @accid.ges required.</sch:assert>
  </sch:rule>
</pattern>
<pattern xmlns="http://purl.oclc.org/dsdl/schematron">
  <!-- following occurrences of diatonic pitch in measure, including grace notes -->
  <sch:rule xmlns:rng="http://relaxng.org/ns/structure/1.0"
    xmlns:sch="http://purl.oclc.org/dsdl/schematron"
    xmlns:a="http://relaxng.org/ns/compatibility/annotations/1.0"
    xmlns="http://www.tei-c.org/ns/1.0" xmlns:math="http://exslt.org/math"
    context="mei:note[@accid and //mei:*[@accid.ges]]">
    <sch:let name="this.pname" value="@pname"/>
    <sch:let name="this.measure" value="generate-id(ancestor::mei:measure)"/>
    <sch:assert role="warning"
      test="not(following::mei:note[@pname=$this.pname and not(@accid) and 
      generate-id(ancestor::mei:measure) eq $this.measure][not(@accid.ges)])"
      >When @accid.ges is used anywhere, notes in the same measure that don't already use
      @accid following a note with a written accidental with the same pitch name should
       provide @accid.ges.</sch:assert>
  </sch:rule>
</pattern>
<pattern xmlns="http://purl.oclc.org/dsdl/schematron">
  <!-- occurrence of pitch in previous measure if tied over barline -->
  <sch:rule xmlns:rng="http://relaxng.org/ns/structure/1.0"
    xmlns:sch="http://purl.oclc.org/dsdl/schematron"
    xmlns:a="http://relaxng.org/ns/compatibility/annotations/1.0"
    xmlns="http://www.tei-c.org/ns/1.0" xmlns:math="http://exslt.org/math"
    context="mei:note[@accid and position()=last() and (@tie='i' or @tie='m') or (@xml:id eq //mei:tie/substring(@startid, 2)) and //mei:*[@accid.ges]]">
    <sch:let name="this.pname" value="@pname"/>
    <sch:let name="this.staff" value="ancestor::mei:staff/@n"/>
    <sch:let name="this.layer" value="ancestor::mei:layer/@n"/>
    <sch:assert role="warning"
      test="ancestor::mei:measure/following-sibling::mei:measure[1]/mei:staff[@n=$this.staff]//mei:note[@pname=$this.pname][1][@accid.ges]"
      >When @accid.ges is used anywhere and the last note of a measure is tied across the bar
      line, the first note in the next measure on the same staff and layer should provide
      @accid.ges.</sch:assert>
  </sch:rule>
</pattern>
```

These rules
1. allow `@accid.ges` to continue to be ignored; that is, existing encodings without it are still valid; but
2. become "active" when `@accid.ges` is used anywhere.

Rules 1 and 2 (for sharp and flat keys, respectively) propose that notes altered by the key signature should carry `@accid.ges`.  In the example below, assuming `@accid.ges` occurs somewhere in the file, the 2nd note (and all F's throughout the file, including grace notes) should be explicitly altered as well.  The rule is "key signature aware", so if the key signature were "3s", then the G's would have to provide `@accid.ges` as well.

```
<score>
  <scoreDef xml:id="scoredef-000000179309736">
    <staffGrp xml:id="m-000000070692809">
      <staffDef xml:id="staffdef-000000146094065" clef.shape="G" clef.line="2" key.sig="1s"
        meter.count="4" meter.unit="4" n="1" lines="5"/>
    </staffGrp>
  </scoreDef>
  <section>
    <measure xml:id="measure-000000184903562" n="1" right="end">
      <staff xml:id="staff-000000187575861" n="1">
        <!-- kern: =1-  4g  4f#  2g  == -->
        <layer xml:id="layer-000000081512115" n="1">
          <note xml:id="note-000000094301745" dur="4" oct="4" pname="g"/>
          <note xml:id="note-000000086498409" dur="4" oct="4" pname="f" accid.ges="s"/>
          <note xml:id="note-000000147091173" dur="2" oct="4" pname="g"/>
        </layer>
      </staff>
    </measure>
  </section>
</score>
```

Rule 3, again assuming the use of `@accid.ges` somewhere in the file, suggests that notes in the same measure with the same pitch name (but not necessarily in the same octave) following a note with a written accidental (encoded using `@accid`) and not already carrying their own accidental should have `@accid.ges`.  For example, the 4th note below should provide `@accid.ges` because the 2nd note has a written accidental.  This rule also applies equally to grace and "regular" notes. 

```
<score>
  <scoreDef xml:id="scoredef-000000074864452">
    <staffGrp xml:id="m-000000046085213">
      <staffDef xml:id="staffdef-000000171571782" clef.shape="G" clef.line="2" key.sig="1s"
        meter.count="4" meter.unit="4" n="1" lines="5"/>
    </staffGrp>
  </scoreDef>
  <section>
    <measure xml:id="measure-000000180660692" n="1" right="end">
      <staff xml:id="staff-000000126160040" n="1">
        <!-- kern: =1-  4g  4f  4g  4f  == -->
        <layer xml:id="layer-000000161192639" n="1">
          <note xml:id="note-000000114062982" dur="4" oct="4" pname="g"/>
          <note xml:id="note-000000212643565" dur="4" accid="n" oct="4" pname="f"/>
          <note xml:id="note-000000058116659" dur="4" oct="4" pname="g"/>
          <note xml:id="note-000000091136922" dur="4" oct="4" pname="f" accid.ges="n"/>
        </layer>
      </staff>
    </measure>
  </section>
</score>
```

(This is probably a good time to note that, while the use `@accid.ges` is encouraged, there is no check for a "correct" value, something that's very difficult to ascertain as it depends on various conventions. For example, grace notes may or may not be affected by the accidentals on regular notes, regular notes are never influenced by the accidentals on grace notes, etc. depending on the notational repertoire/style.)

Rule 4 encourages the use of `@accid.ges` for the the terminating note of an across-the-barline tie encoded using either `@tie` attributes or a `<tie>` element.

```
<score>
  <scoreDef xml:id="scoredef-000000080124437">
    <staffGrp xml:id="m-000000174655344">
      <staffDef xml:id="staffdef-000000036977927" clef.shape="G" clef.line="2" key.sig="1s"
        meter.count="4" meter.unit="4" n="1" lines="5"/>
    </staffGrp>
  </scoreDef>
  <section>
    <measure xml:id="measure-000000006255008" n="1">
      <staff xml:id="staff-000000111264048" n="1">
        <!-- kern: =1-  2g  [2fn  = -->
        <layer xml:id="layer-000000200856738" n="1">
          <note xml:id="note-000000169662611" dur="2" oct="4" pname="g"/>
          <note xml:id="note-000000090726674" dur="2" accid="n" oct="4" pname="f"/>
        </layer>
      </staff>
      <tie xml:id="tie-000000148794532" startid="#note-000000090726674"
        endid="#note-000000139013283"/>
    </measure>
    <measure xml:id="measure-000000129830632" right="end">
      <staff xml:id="staff-000000104516787" n="1">
        <!-- kern: =  2f]  2f#  == -->
        <layer xml:id="layer-000000186774311" n="1">
          <note xml:id="note-000000139013283" dur="2" oct="4" pname="f" accid.ges="n"/>
          <note xml:id="note-000000045000796" dur="2" accid.ges="s" oct="4" pname="f"/>
        </layer>
      </staff>
    </measure>
  </section>
</score>
```

Currently, these rules provide only warnings, but in a future release they could be made to raise validation errors.  In any case, if we wish to maintain backward compatibility, these rules should be placed in a new module, named something like "MEI.cmn.strict".  In addition to these rules, this module could contain other requirements necessary for strict adherence to rules expected for certain uses, e.g., within a notation editor that expects certain markup to be present in order to provide particular functionality.  Since these rules won't be "baked into" MEI, of course users who expect these behaviors will have to remember to activate the MEI.cmn.strict module.  But, that's no different from expecting users to intentionally invoke/ignore other modules.

(If the idea of a required-yet-optional module troubles anyone, I'd like to draw your attention to the MEI and MEI.shared modules, which are optional in the sense that they need not be invoked by a customization, but are required in the sense that if they are removed, then MEI will cease to function.)

The modular approach suggests the creation of additional "rule sets" for other situations -- specific rules for neume, black/white mensural, CMN, or post-modern notation, for example.  For this to work, however, the base schema must remain as flexible as possible -- "half-baked" as it has been called.  (Just for the record, the technical definition of "flexibility" is the capacity to work for a wide variety of uses, even unforeseen ones.)

Using Schematron provides another benefit over the creation of required elements/attributes; that is, it allows one to provide whatever granularity is necessary for any given markup structure.  The "all-or-nothing" kind of encoding mentioned earlier can also be avoided using this approach.  For example, these rules address only the use of `@accid.ges`, without saying anything about the handling of other attributes (gestural or otherwise) alone.  Additional rules can be added incrementally for those as well, but we **need not alter the schema to do it**.  The problem with altering the schema, of course, is that often what seems like a good idea today may cause issues tomorrow.

**Comment by pe-ro** [comment]
If `@accid.ges` is required in order to produce MIDI output, then why not parse the file against a purpose-built schema that requires it before starting the MIDI creation process?  If the file is invalid, then the processor can refuse to generate MIDI output or issue a warning that the MIDI will be fubar.

Alternatively, XPath could be used to compare the number of notes in the file with the number of notes having `@accid.ges`.  If these values aren't the same, then the processor's options are the same as above.

I'm aware that these alternatives require calculation (and therefore time), but they are also much simpler than what's been suggested so far.

**Comment by kepper** [comment]
Thanks for your comments, @pe-ro. Your Schematron rules seem to be a good approach, but I understand them more as a complementary solution than a rivalling one. As soon as you don't trigger them by the presence of a single `@accid.ges` somewhere in the file, but instead by the value provided in `@gestural` (or whatever the name for this single additional attribute will be), they perfectly fit into @lpugin's and my proposal (which is really just a compilation of arguments brought to the table by each of us). I'm sure there are certain things to discuss about these Schematron rules, but let's not go there for the moment. Let's first decide if we want to go in the direction Laurent and I suggested, and only as soon as we have an agreement on the direction, let's work out the implementation details. 

Re: your last comment: I think we're all ok with calculating things as necessary. All we'd like to have is a simple indication if calculations _are_ necessary or not (or, in order to allow a gentle transition into this distinction, if we _explicitly_ don't know if it's necessary and have to fall back to a mechanism of our own choice). 

So are you (and everyone else here as well…) cool with the idea to have one general attribute adressing all things gestural in one shot, and asking on MEI-L if there is need for an additional finer-grained mechanism on top of that?

**Comment by pe-ro** [comment]
> All we'd like to have is a simple indication if calculations are necessary or not (or, in order to allow a 
> gentle transition into this distinction, if we explicitly don't know if it's necessary and have to fall back
> to a mechanism of our own choice).
> 
> So are you (and everyone else here as well…) cool with the idea to have one general attribute 
> adressing all things gestural in one shot, and asking on MEI-L if there is need for an additional finer-
> grained mechanism on top of that?

Honoring the long-standing tradition of distrusting user input, I think having one general, **required** attribute that addresses all things gestural is virtually useless without a finer-grained mechanism.  For example, what happens if I say `@gestural="true"` but don't actually use any gestural domain attributes like `@accid.ges` or vice versa?  In order to guard against this, the fine-grained details must also be required.  At that point, the processor-dependent markup camel has gotten its nose under the tent and the whole idea of software/processor agnosticism is compromised.  If you're planning to go this way, I don't think a poll is necessary as the outcome is pre-determined.

I prefer an _informative_ rather than a normative approach; that is, using a combination of the Schematron rules I proposed yesterday and optional markup in `<encodingDesc>` similar to TEI's `<tagsDecl>`, the processor can be more reliably informed about the markup.  It is true, however, that when none of this information is present, the processor must provide its own fall-back procedures.  But that's also true of any file bearing `@gestural="false"` or `@gestural="undetermined"`.  I suppose the discussion hinges on what a processor would likely do when `@gestural` has a value other than "true".

**Comment by pe-ro** [comment]
As I suggested earlier today, the simplest approach is for the processor to determine if the information it requires is present; that is, not using the "second hand news" value of `@gestural`, but by actually examining the file's content.  This can be accomplished using a more strict schema or through some calculation, e.g. XPath.

**Comment by lpugin** [comment]
> what happens if I say @gestural="true" but don't actually use any gestural domain attributes like @accid.ges 

This is no problem and actually the reason why we need this. For example, if  you have a piece is C-major with no modulation, there will be no `@accid.ges` but `@gestural="true"` indicates that you can render it in sound without doing any processing.

> or vice versa?

Here we can provide a Schematron rule that says that something is wrong. But it will not tell you what.  What you are trying to do with your set of Schematron rules is to validate music notation rules of the various way of writing / interpreting accidentals. This can be useful, but the requirement of `@accid.ges` does something different, which is to provide information whereas the encoding of gestural domain has been done or not. If people what to encode junk data, they are always free to do it.

**Comment by lpugin** [comment]
If the problem is the requirement, I think this whole idea of MEI being an encoding framework and not a file format or a standard is going dangerously far when it comes not to require minimal encoding status for information that is of the level of MIDI. As I said, if we do not require something there I fear that this going to be counter-productive for MEI in the long run, not for projects that produce and use their own data, but for projects that might be interested to use existing dataset at an analytical level. (This is not the primarily goal of MEI but something it would not be bad to foresee.)

**Comment by pe-ro** [comment]
OK, I'll concede the need for this and support your proposal for an attribute (required by Schematron) on `<scoreDef>`.  I don't think the name "gestural" is informative enough of the attribute's purpose. However, the best replacement I've been able to come up with so far is "encodedges" or "gesturalenc", neither of which is particularly satisfying, but which attempt to communicate that it's saying something about the encoding process.

I think your proposal needs to be simplified with regard to the recording of finer-grained details.  I suggest using the value "mixed" to be a trigger that requires the use of `<encodingDesc>/<tagsDecl>` instead of other attributes, like `@gestural.events`, `@gestural.trans`, `@gestural.controlEvents`, etc.  `<encodingDesc>/<tagsDecl>` will record information about the use of elements and their attributes on an element-by-element basis.  This will provide a higher level of granularity and flexibility than the extra attributes on `<scoreDef>`. The details regarding `<tagsDecl>` can be worked out later, though.

Frankly, I don't think a poll is necessary, but won't stand in the way.

**Comment by codaich** [comment]
If I might humbly add to this very interesting and insightful thread, which has been contributed to by many who know MEI much better than I, I would like to say that I have developed my own concerns similar to those of @lpugin in many ways, and I think that it is important that certain fundamental issues be emphasized.

I have been doing SIMSSA work for the past little while on automatic statistical feature extraction from MEI encodings, for the purposes of ultimately allowing researchers to search, apply data visualization tools and use machine learning to process huge collections of musical data. Trying to deal with encodings where pitch (and rhythm as well) are unspecified or ambiguous makes it very difficult to do this, as one has no choice but to either 1) treat the MEI file as bad data (and thus discard the entire encoding) or 2) assume a default (which, in the absence of official MEI set of defaults, leaves the maker of each system to assume their own individual defaults).

The first approach is the most empirically responsible, but results in the loss of a lot of otherwise very nice and valuable MEI encodings. In our trial experiments so far, we have certainly found much more than a few MEI files out there in the wild that contain important unresolvable ambiguities with respect to both pitch or rhythm.

The second approach is, I think, a very bad idea because it will either result in completely and arbitrarily different interpretations or renderings of the same MEI encoding by each software system, or it will result in a de facto set of default interpretations being adopted not because they are good or carefully thought out necessarily, but because they are the assumptions adopted (arbitrarily, by necessity) by the most first or most popular existing software.

From a big data perspective, a related additional concern is that certain patterns will emerge in the statistical profiling of MEI encodings not just based on the music or other cultural content they represent (which is desirable), but because of encoding regularities resulting from the particular MEI encoders (human or software), not the actual content that the person (or OMR system, automatic transcriber, etc.) doing the encoding was trying to encode. This could result in data bias that will potentially have very long-term and damaging effects on our ability to apply big data techniques to MEI encodings that contain undesirable ambiguity.

Of course, I do specify “undesirable” ambiguity here; I understand completely that in some cases that it is necessary or desirable to leave certain aspects of an encoding ambiguous, and I think that this should certainly be permitted when appropriate. However, I strongly suspect that the significant majority of MEI encodings with ambiguity that are out there have ambiguity due to error (human or software), rather than measured scholarly consideration. So, as others have suggested before, I think that any MEI encoding that contains certain important kinds of ambiguity (e.g. pitch and rhythm) must be flagged as such, and those making encodings should be made explicitly aware that if they decide to maintain this, it may mean that their work may not be properly rendered, and may be discarded by analysis and indexing tools.

As a fantastic framework in many ways, I think MEI has the potential to be hugely useful for a great variety of purposes. Having said that, for it to continue to grow at its current impressive rate, I think the focus may need to be shifted a little more from primarily the perspectives of those doing encodings to also including a little more emphasis on the priorities of those who will ultimtately be using the encodings, or using software to interpret them. This is true in a range of areas, including automatic analysis and indexing, as I’m doing, as well as for visualizing the data as digital manuscripts, as @lpugin is doing, or even for sonifying musicy. Ultimately, I think that if we fail to make sure that the very significant majority of MEI encodings produced are consistent and complete with respect to essential information when it is possible for them to be so, then we will ultimately be limiting the usefulness of the data so many people are putting such great work into producing.

As to the nuts and bolts of the best MEI standards that should be adopted in order to achieve this, I will leave this to the many others who have much greater MEI wizardry than myself. We are happy to write code for interpreting whatever the community agrees on, just as long as all the information that we need to meaningfully interpret is actually there in the encoding.

**Comment by kepper** [comment]
:-)

We discussed this briefly before I had to leave, and we also came to the conclusion that it
- might be better to include the finer-grained approach right away
- explore if we shouldn't do this with something like tagsDecl instead of attributes

So eventually, we ended on the same page. Probably that's a good indication that this approach isn't too bad ;-)

Even though we technically don't need the poll, I'd be curious about its results, so I'd like to do that anyway, but probably not before next week. We can start to work on the implementation details like names etc then as well. Sound good?

**Comment by pe-ro** [comment]
@kepper: Sure.

**Comment by pe-ro** [comment]
Just for fun :-), here's a head start on a possible approach to the details of expanding `<encodingDesc>`:

```
in modified DTD syntax:

encodingDesc (appInfo?, editorialDecl?, projectDesc?, samplingDecl?, tagsDecl?)

tagsDecl (namespace)+

namespace (tagUsage)+

tagUsage (desc?, attUsage*)
   attr name = element name, required
   attr context = circumstances of the element's use, required
   attr occurs = number of occurrences in the defined context, optional
   attr withid = number of occurrences in the defined context with @xml:id, optional

attUsage (desc?)
   attr name = attribute name, required
   attr context =  circumstances of the attribute's use, optional/required?
```

`<tagsDecl>` is optional, but once it's used, it requires content.  `<namespace>` is necessary because MEI already allows elements from several namespaces and could add others in the future.  A human-readable description of the use of an element in a given context can be captured in `<desc>`.  An element may have multiple entries (in the same namespace) in order to record different contexts in which the element can appear.  It will be best if the value of `@context` is machine-readable; that is, XPath.  Similarly, each attribute used by the element can be recorded in `<attUsage>`.  I'm not sure at this point about `attUsage/@context` -- its usefulness depends on whether the attribute usage context is tied to / inherited from the element usage context.  This will require some more thought.

The `<tagsDecl>` element isn't intended to address the content of elements, in either their "normal" or customized state.  The place for that is in an `<elementSpec>` in mei-source.xml, a customization, or (if we get around to allowing it) a `<schemaSpec>` within `<encodingDesc>`.

`<tagsDecl>` can also be used to document so-called _foreign_ XML provided in the `<extMeta>` element.

**Comment by lpugin** [comment]
`scoreDef` attribute: I have `@encodedges` (or `@encoded.ges`?) better than `gesturalenc`.

**Comment by pe-ro** [comment]
The `.ges` suffix was intended to identify optional, gestural attributes so they could easily be "turned off".  This is something more "meta" -- whether those attributes are present or not -- so I prefer `@encodedges`.

**Comment by lpugin** [comment]
OK, I understand. I am fine with `@encodedges`.

**Comment by pe-ro** [comment]
I just realized that requiring `@encodedges` on `<scoreDef>` means that it would have to appear anytime `<scoreDef>` was used to reset score-level attributes.  This probably isn't acceptable, so I think we'll have to find a place higher up in the hierarchy to place this attribute.  We could require it on `<score>` and `<part>` instead.  Another option would be to make it an optional attribute on scoreDef, score, part, mdiv, body, music, and mei, but add a rule on scoreDef that says it has to be present on this element or one of the ancestors I just listed.

I think I'd like to also reconsider the name of this attribute.  I'm thinking `@ges` or `@uses.ges` will be simple and easy to remember (both the name and the attribute's function).

**Comment by pe-ro** [comment]
The following example illustrates how to document elements in the MEI and SVG namespaces, and attributes in the XML namespace using `<tagsDecl>` --

```
<tagsDecl>
  <desc></desc>
  <namespace name="http://www.music-encoding.org/ns/mei" prefix="mei">
    <tagUsage name="measure">
      <desc></desc>
      <attUsage name="n">
        <desc></desc>
      </attUsage>
    </tagUsage>
  </namespace>
  <namespace name="http://www.w3.org/2000/svg" prefix="svg">
    <tagUsage name="svg" context="mei:surface" occurs="1"/>
    <tagUsage name="svg" context="mei:zone" occurs="1"/>
    <tagUsage name="svg" context="mei:fig" occurs="1"/>
    <tagUsage name="svg" context="mei:symbolDef" occurs="1"/>
  </namespace>
  <namespace name="http://www.w3.org/XML/1998/namespace" prefix="xml">
    <attUsage name="base" context="mei:mei"/>
    <attUsage name="id"/>
    <attUsage name="lang"/>
    <attUsage name="space"/>
  </namespace>
</tagsDecl>
```
Elements/attributes always belong to a namespace (even if it's the no-namespace namespace), so `<namespace>` is the organizing factor within `<tagsDecl>`.  `<desc>` is provided at each level to accommodate human-readable explanations, while the children of `<namespace>` are intended to carry machine-processable info.  

Inside `<namespace>` one can either choose to record elements/tags or attributes, but not both.  If elements are chosen; that is, if `<tagUsage>` appears, then the attributes associated with that element are described using `<attUsage>` children of `<tagUsage>`.  Otherwise, `<attUsage>` elements appear directly within `<namespace>`.  This is intended to allow for the situation where there are no elements in the namespace, like in XML.  

Occurrences of elements can be described in terms of context and/or number.  `@context` holds an XPath statement, while `@occurs` records a count of elements *in that context*.  `@withid` holds the number of elements in that context that have an xml:id.

`<attUsage>` also uses `@context` to record where the attribute occurs, but it doesn't have `@occurs`, since an attribute can only occur once on a given element.

If an element is used in more than one context, either `<tagUsage>` can be repeated (as above) or a more general XPath can be employed, for example --

```
<tagUsage name="svg"
    context="mei:*[local-name() eq 'surface' or local-name() eq 'zone' or 
    local-name() eq 'fig' or local-name() eq 'symbolDef']" occurs="4"/>
```

**Comment by pe-ro** [comment]
Assuming that we're agreed on the issue of providing element-by-element encoding description, I've already added `<tagsDecl>` as an optional element within `<encodingDesc>`, which is also optional.

I'd like to move on to the question of adding an indication of the presence/absence of domain-specific encoding.  Having thought about this a little more, I think we can proceed with recommending the use of `@ges` on `<score>` and `<parts>`.  Marking the attribute as "recommended" in the ODD means it will still be optional, but this gives us a documented reason to make it required in the next-next release.  

Another option, however, is to put this in the header, as someone suggested before -- Sorry, sometimes I'm a little slow, but I eventually get there.  :)  Since making an indication of the encoding of the various domains *mandatory* within the *optional* `<encodingDesc>` element won't actually achieve what we're looking for, how about adding a `<domainsDecl>` element as a sibling of `<encodingDesc>`?

`<domainsDecl>` could have required attributes `@anl`, `@ges`, `@vis`, and `@log` (just for completeness' sake).  The element can be optional in the next release, and required in the next-next.

Whaddaya think?

**Comment by pe-ro** [comment]
Having heard no further comments, I suppose my last suggestion is ok with everyone here.  However, I'm still not sure about it.  At this point, I'd rather put `<domainsDecl>` inside `<encodingDesc>`, but I'm reluctant to require `<encodingDesc>`.  This would break every existing MEI file.  The upgrade XSLT can supply "best guess" data by looking at the usage of analytical, gestural, and visual attributes, but every file would still have to be examined for correctness.  So, I ask once again, Is it absolutely necessary to require domain usage?  If so, can we live with ` <domainsDecl>` and `<encodingDesc>` being optional in the short run and requiring them at some point in the future?

**Comment by pe-ro** [comment]
No comments?

**Comment by kepper** [comment]
> Am 22.02.2017 um 17:20 schrieb pe-ro <notifications@github.com>:
> 
> Is it absolutely necessary to require domain usage?

I could live without this, but…

> If so, can we live with <domainsDecl> and <encodingDesc> being optional in the short run and requiring them at some point in the future?
> 
> 

this solution seems sensible to me. Make it optional in the upcoming release, recommended in the one after that, and required thereafter… (one step less in the procedure would be equally fine)

**Comment by pe-ro** [comment]
Here's the chunk of ODD that implements `<domainsDecl>` --
```
<elementSpec ident="encodingDesc" module="MEI.header" mode="replace">
  <desc>(encoding description) – Documents the relationship between an electronic file and
    the source or sources from which it was derived as well as applications used in the
    encoding/editing process.</desc>
  <classes>
    <memberOf key="att.bibl"/>
    <memberOf key="att.common"/>
    <memberOf key="model.headerPart"/>
  </classes>
  <content>
    <rng:optional>
      <rng:ref name="appInfo"/>
    </rng:optional>
    <rng:optional>
      <rng:ref name="editorialDecl"/>
    </rng:optional>
    <rng:optional>
      <rng:ref name="projectDesc"/>
    </rng:optional>
    <rng:optional>
      <rng:ref name="samplingDecl"/>
    </rng:optional>
    <rng:zeroOrMore>
      <rng:ref name="domainsDecl"/>
    </rng:zeroOrMore>
    <rng:optional>
      <rng:ref name="tagsDecl"/>
    </rng:optional>
  </content>
  <remarks>
    <p>This element is modelled on an element in the Text Encoding Initiative (TEI)
      standard.</p>
  </remarks>
</elementSpec>
<elementSpec ident="domainsDecl" module="MEI.header" mode="add">
  <desc>(domains declaration) – Indicates which domains are included in the encoding.</desc>
  <classes>
    <memberOf key="att.bibl"/>
    <memberOf key="att.common"/>
  </classes>
  <attList>
    <attDef ident="anl" usage="req">
      <datatype>
        <rng:ref name="data.BOOLEAN"/>
      </datatype>
    </attDef>
    <attDef ident="ges" usage="req">
      <datatype>
        <rng:ref name="data.BOOLEAN"/>
      </datatype>
    </attDef>
    <attDef ident="vis" usage="req">
      <datatype>
        <rng:ref name="data.BOOLEAN"/>
      </datatype>
    </attDef>
  </attList>
</elementSpec>
```
`<domainsDecl>` is optional (as discussed above).  It's also repeatable so that multiple combinations of `@anl`, `@ges`, and `@vis` may be defined, then pointed to by different parts of the encoding. Different movements, staves, layers, or even individual elements may be associated with a particular `<domainsDecl>` element via `@decls`.
